test_functional.py

```diff
@@ -4,62 +4,44 @@
 class TestRomanToDecimal(unittest.TestCase):
 
     def test_sanity(self):
-        self.assertEqual(roman_nums.rom_num_func('I'), 1)
+        self.assertEqual(roman_nums.convert('I'), 1)
+
+    def test_error_cases(self):
+        error_cases = ["IXI","VV","IIII","IM","VXV"]
+        for case in error_cases:
+            with self.assertRaises(ValueError):
+                roman_nums.convert(case)
+
+    def test_specific_cases(self):
+        test_cases = [("MDCLXVI",1666), ("XIV",14), ("     i      X     ",9)]
+        for roman, decimal in test_cases:
+            self.assertEqual(roman_nums.convert(roman), decimal)
 
-    def test_thorough(self):
-        test_one = "MDCLXVI"
-        self.assertTrue(roman_nums.valid_entry_f(test_one))
-        self.assertEqual(roman_nums.rom_num_func(test_one), 1666)
-        test_two = "IXI"
-        self.assertFalse(roman_nums.valid_entry_f(test_two))
-        test_three = "VV"
-        self.assertFalse(roman_nums.valid_entry_f(test_three))
-        test_four = "XIV"
-        self.assertTrue(roman_nums.valid_entry_f(test_four))
-        self.assertEqual(roman_nums.rom_num_func(test_four), 14)
-        test_five = "IIII"
-        self.assertFalse(roman_nums.valid_entry_f(test_five))
-        test_six = "     i      X     " 
-        self.assertTrue(roman_nums.valid_entry_f(test_six))
-        self.assertEqual(roman_nums.rom_num_func(test_six), 9)
-        test_seven = "IM"
-        self.assertFalse(roman_nums.valid_entry_f(test_seven)) 
-        test_eight = "VXV"
-        self.assertFalse(roman_nums.valid_entry_f(test_eight))
 
 class TestDecimalToRoman(unittest.TestCase):
 
     def test_sanity(self):
         self.assertEqual(roman_nums.numerical(1), 'I')
 
-    def test_thorough(self):
-        test_one = 4
-        self.assertTrue(roman_nums.valid_entry_f(test_one))
-        self.assertEqual(roman_nums.numerical(test_one), "IV")
-        test_two = 999
-        self.assertTrue(roman_nums.valid_entry_f(test_two))
-        self.assertEqual(roman_nums.numerical(test_two), "CMXCIX")
-        test_three = 1342
-        self.assertTrue(roman_nums.valid_entry_f(test_three))
-        self.assertEqual(roman_nums.numerical(test_three), "MCCCXLII")
-        test_four = "   2   "
-        self.assertTrue(roman_nums.valid_entry_f(test_four))
-        self.assertEqual(roman_nums.numerical(test_four), "II")
-        test_five= "    2  V"
-        self.assertFalse(roman_nums.valid_entry_f(test_five)) 
-        test_six = "4000"
-        self.assertFalse(roman_nums.valid_entry_f(test_six))
+    def test_error_cases(self):
+        error_cases = ["    2  V","-1","0","4000"]
+        for case in error_cases:
+            with self.assertRaises(ValueError):
```

Comment:
What does this code chunk mean, in specific the "with" statement

Reply:
Good question! "with" is a python built-in that sets up a context. I can't explain it as well as the docs: https://docs.python.org/3/reference/compound_stmts.html#the-with-statement

And in this use I'm following what the unittest library and assertRaises show for an example, here: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises